lib/web_ui/lib/src/engine/view_embedder/display_dpr_stream.dart

```diff
@@ -0,0 +1,93 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+import 'dart:js_interop';
+
+import 'package:meta/meta.dart';
+import 'package:ui/src/engine/display.dart';
+import 'package:ui/src/engine/dom.dart';
+import 'package:ui/ui.dart' as ui show Display;
+
+/// Provides a stream of `devicePixelRatio` changes for the given display.
+///
+/// Note that until the Window Management API is generally available, this class
+/// only monitors the global `devicePixelRatio` property, provided by the default
+/// [EngineFlutterDisplay.instance].
+///
+/// See: https://developer.mozilla.org/en-US/docs/Web/API/Window_Management_API
+class DisplayDprStream {
+  DisplayDprStream(
+    this._display, {
+    @visibleForTesting DebugDisplayDprStreamOverrides? overrides,
+  })  : _currentDpr = _display.devicePixelRatio,
+        _debugOverrides = overrides {
+    // Start listening to DPR changes.
+    _subscribeToMediaQuery();
+  }
+
+  /// A singleton instance of DisplayDprStream.
+  static DisplayDprStream instance =
+      DisplayDprStream(EngineFlutterDisplay.instance);
```

Comment:
Do we want to continue with the singleton pattern? I was under the impression that we were removing all singletons except one top-level class (was it `EnginePlatformDispatcher`?), which would do a kind of dependency injection.

Reply:
Yeah, we're still using the singleton pattern for things that are ONE only (browser globals). `devicePixelRatio` is determined by the browser window. This is similar to the "brightness/dark mode" detector, and implemented in an almost identical way.

_(I tried to make this instance-based (by passing a Display), hoping that this API[^1] will be eventually usable, but it's still unclear to me from the docs how it can be used to monitor DPR changes, though!)_

[^1]: https://developer.mozilla.org/en-US/docs/Web/API/Window_Management_API

Reply:
Ah, I understand now! Yes, I could move this to a singleton of the EnginePlatformDispatcher instead of being a loose object.

Reply:
I think the litmus test is that code that consumes an external service should not go through a static to get a reference to that service. `EnginePlatformDispatcher.instance.displayDprStream` still accesses `DisplayDprStream` through a static getter. In fact, it may be worse, the code only needed a `DisplayDprStream` but it depends on `EnginePlatformDispatcher`, which vends more dependencies than it needs. It still has the same testability issues, where to mock out `DisplayDprStream` you need to mock out the `EnginePlatformDispatcher.instance` static getter.

I don't insist that this is fixed in this PR, unless we already switched modes to a singleton-free design.

Reply:
I changed this a little bit. The CustomElement dimensions provider now just requires a Stream<double> passed by its constructor. For now, the abstract factory class of DimensionsProviders knows where to get that stream from (the singleton instance), but it's very likely that it could just be created from there.

Anyway, I'm going to `autosubmit` this, and if you find anything egregious, feel free to raise an issue or ping me so I can address it ASAP.